### PR TITLE
Add persistent Notebook Mode and true Codex Fast Mode

### LINKED ISSUE #241: [pi-codex-conversion]: Add Notebook Code Mode
## Package

pi-codex-conversion

## Goal

Add Notebook Code Mode alongside normal and existing V8 Code Mode. Preserve the provider-visible `exec` and `wait` contract and existing nested tools while running sequential JavaScript/TypeScript cells in one persistent Deno Jupyter kernel.

## Scope

- Linux x64 internal proof first
- session and trusted-project mode selection
- persistent declarations, imports, npm packages, and top-level await
- existing Pi/custom-tool bridge, deferred discovery, and output behavior
- busy/yield/wait/cancellation handling
- bounded best-effort checkpoint and recovery
- no changes to the existing V8 Code Mode runtime
- no Prime harness, memory, skills, or subagent bundle
- no Cloudflare durable-runtime dependency; its replay log is orthogonal to notebook heap persistence

The implementation plan is in `docs/pi-execution-modes.md`.

## Completion

The proof satisfies the plan's prototype acceptance criteria, includes a package changeset, and passes the changed-package gate.

### LINKED ISSUE #284: Support credential-scoped aliases for canonical Codex providers
## Summary

Canonical Codex subscription behavior is currently coupled in several places to the registry provider ID `openai-codex`.

Provider IDs also act as credential scopes in Pi, so a provider using the same canonical Codex API contract and ChatGPT backend can lose Codex subscription capabilities solely because it uses a distinct provider ID for an isolated credential scope.

A representative case is:

```text
provider = openai-codex-personal
api      = openai-codex-responses
baseUrl  = https://chatgpt.com/backend-api
```

The intended behavior is for that model to keep Codex transport/subscription capabilities while resolving authentication from `openai-codex-personal`, not from the stock `openai-codex` credential scope.

## Proposed behavior

The distinction that seems to work cleanly is:

- `model.provider` defines the credential scope;
- the stock `openai-codex` provider or a canonical Codex API/backend contract defines Codex transport behavior;
- subscription-only APIs remain restricted to canonical ChatGPT endpoints and canonical subscription auth.

This preserves existing stock `openai-codex` behavior, including custom endpoints, while allowing canonical provider aliases to use their own credentials.

Generic Responses providers would not gain automatic Codex subscription access, and explicitly configured proxy behavior would remain unchanged.

## Patch

I have a tested patch prepared here:

https://github.com/rkbkosp/pi-codex-conversion/commit/4dd6dc94d1b6388bbb27a62deb055e98d96c5ea5

It covers:

- alias-specific credential resolution for usage and reset credits;
- canonical subscription vs Codex transport separation;
- Code Mode and Responses Lite behavior;
- native compaction;
- Codex-backed tool routing;
- tool-history serialization;
- regression coverage for stock `openai-codex` custom endpoints;
- rejection of non-canonical aliases for subscription-only APIs.

I tried to open this as a draft PR, but the repository currently limits PR creation to collaborators. If this direction is acceptable, I’d be happy to submit it in whatever form you prefer.

### LINKED ISSUE #288: [pi-codex-conversion]:  Pi's shellPath configuration setting is being honored
### Which package is this about?

pi-codex-conversion

### Skill name, if this is about a skill

_No response_

### What kind of thing is it?

Bug / broken behavior

### What happened?

Hey, huge fan of the project here. 

Thanks for making pi-codex-conversion, it's been amazing so far. Ok, onto the bug...

Disclosure: Clanker authored report, human reviewed.
---
On Windows, I configured Pi to use MSYS2 Bash:

 ```json
   {
     "shellPath": "C:\\msys64\\usr\\bin\\bash.exe"
   }
 ```

 Pi’s native shell uses that configuration correctly. However, pi-codex-conversion’s exec_command defaults to Git Bash instead:

 ```text
   C:\Program Files\Git\bin\bash.exe
 ```

 This causes commands to run in a different MSYS installation with a different /usr/bin and toolset. For example, tmux is available in my configured
 MSYS2 environment but not in Git Bash:

 ```sh
   command -v tmux
   tmux -V
 ```

 Actual result through the default exec_command shell:

 ```text
   tmux: command not found
 ```

 Expected result:

 ```text
   /usr/bin/tmux
   tmux 3.7b
 ```

 Passing the configured shell explicitly to exec_command works, so the issue appears limited to default-shell resolution:

 ```json
   {
     "cmd": "command -v tmux; tmux -V",
     "shell": "C:\\msys64\\usr\\bin\\bash.exe"
   }
 ```

### Quick repro, if you have one

_No response_

### Pi version

0.84.1

### Anything else?

 Environment:

 - Windows
 - Pi Coding Agent: 0.84.1
 - @howaboua/pi-codex-conversion: 3.0.14
 - Configured shell: C:\msys64\usr\bin\bash.exe
 - Incorrectly selected shell: C:\Program Files\Git\bin\bash.exe

 Relevant Pi setting:

 ```json
   {
     "shellPath": "C:\\msys64\\usr\\bin\\bash.exe"
   }
 ```

 My ~/.pi/agent/pi-codex-conversion.json configuration:

 ```json
   {
     "mode": "normal",
     "voiceFeaturesOnly": false,
     "prompt": {
       "heavySystemPromptOverwrite": true
     },
     "scope": {
       "allProviders": "off",
       "additionalProviders": []
     },
     "tools": {
       "customRustBinariesDir": "",
       "webRun": true,
       "imageGeneration": true,
       "viewImageFallback": false,
       "applyPatchOnly": false,
       "viewImageOnly": false,
       "webRunOnly": false,
       "imageGenerationOnly": false
     },
     "ui": {
       "statusLine": true,
       "toolRenaming": true,
       "compactTools": false,
       "codeModeDetails": true,
       "backgroundShellWidget": true,
       "backgroundShellToggleShortcut": "alt+w",
       "backgroundShellPrevShortcut": "alt+q",
       "backgroundShellNextShortcut": "alt+e",
       "backgroundShellCloseShortcut": "alt+r"
     },
     "compaction": {
       "responsesCompaction": true,
       "version": "v1"
     },
     "beta": {
       "codeMode": true,
       "responsesLite": false,
       "v2UserMessageRetention": 64
     },
     "voice": {
       "protocol": "v3",
       "mode": "conversational",
       "v2Voice": "marin",
       "v3Voice": "cove",
       "dictationShortcut": "ctrl+alt+d",
       "realtimeShortcut": "ctrl+alt+space",
       "dictationShortcutMode": "push",
       "autoResumeRealtime": false,
       "muteShortcut": "ctrl+alt+m",
       "serverShortcut": "ctrl+alt+g",
       "contextReasoning": "high"
     },
     "openai": {
       "fast": false,
       "verbosity": "low",
       "forceCachedWebSockets": true,
       "webSearchModel": "gpt-5.6-luna",
       "cacheDiagnostics": "off",
       "harnessIdentifierHeader": false
     }
   }
 ```

 The likely cause is getDefaultCodexRuntimeShell() in src/adapter/prompt/runtime-shell.ts:

 ```ts
   export function getDefaultCodexRuntimeShell(): string {
     if (process.platform === "win32") {
       return getShellConfig().shell;
     }
     return getCodexRuntimeShell(process.env["SHELL"]);
   }
 ```

 getShellConfig() is called without Pi’s configured shellPath. On Windows, it consequently follows automatic detection and prefers Git Bash. Pi’s native
 shell tooling reads settingsManager.getShellPath() and passes it explicitly.

 Providing the shell directly in an exec_command call is a successful workaround:

 ```json
   {
     "cmd": "command -v tmux; tmux -V",
     "shell": "C:\\msys64\\usr\\bin\\bash.exe"
   }
 ```

 This returns:

 ```text
   /usr/bin/tmux
   tmux 3.7b
 ```

### PR DESCRIPTION
**Pull request by howaclawa**
This PR adds an opt-in persistent Deno/TypeScript Notebook runtime to Code Mode, activates the backend's real Fast Mode contract, and carries the Codex alias, shell, and voice fixes already reviewed here.

## Summary

- keep Code Mode's `exec`/`wait` workflow while making each `exec` the next cell in a persistent Jupyter-style kernel
- retain session scratch privately, promote deliberate `globalThis` values across project sessions, and provide named profiles plus `status`, pinning, pruning, diagnostics, restart, and reset lifecycle controls
- checkpoint at lifecycle boundaries, maintain recoverable append-oriented `.ipynb` journals, isolate concurrent live sessions, report memory pressure, and expose nested Notebook tool activity
- download and verify pinned Deno 2.9.5 lazily for Linux, macOS, and Windows on x64 and ARM64 rather than bundling a runtime
- activate true ChatGPT Codex Fast Mode only when `openai.fast` produces `service_tier: priority`, using `originator: codex_cli_rs` and the priority routing hint across prewarm, reconnect, Responses, Responses Lite, WebSocket, SSE/HTTP, fallback, retry, and native compaction paths
- preserve normal request identity when Fast Mode is off while supporting renamed providers, multiple ChatGPT accounts, and monitoring proxies through the shared ChatGPT Codex transport
- let trusted folders snapshot and own the complete `/codex` configuration from the first settings row, while folders without overrides inherit globals and independently launched workers can force Fast Mode per process
- preserve credential-scoped canonical provider aliases, Pi's configured shell behavior, safer nested shell guidance, and bounded realtime voice framing
- expose gated Daybreak Blue and Daybreak Red models through the current Pi catalog, group stock, renamed, and configured proxy Responses Lite tools under the standard `functions` namespace, preserve qualified nested tool identities and end-turn diagnostics, honor Pi strict tool schemas, and require Pi 0.84.2 or newer
- update Undici to patched 8.10.0 and retain manual six-platform Notebook smoke coverage

## Validation

- `bun run check:changed`
- 95 pi-codex-conversion tests and 6 pi-gippity-control tests pass
- extension artifact and package boundaries verified; Knip clean
- deterministic request contracts cover Fast-off isolation, identical Fast routing across WebSocket and HTTP headers, and Codex namespace construction/replay
- live ChatGPT probes accepted default namespace requests and returned non-default namespace identities over WebSocket and SSE
- six-platform kernel, persistent-cell, diagnostics, reset, and shutdown smoke: https://github.com/IgorWarzocha/howaboua-pi-stuff/actions/runs/31255309685

## Review focus

- Notebook state is a private live fork: only deliberate project bindings merge durably; another active kernel is never mutated
- project artifacts remain files; retained Notebook state is compact serializable working memory, not a replacement for source control
- Fast Mode changes transport routing only after the final request confirms priority service; no model-visible prompt or tool prefix changes
- the backend currently reports `response.service_tier: default` even when priority throughput is active, so request intent remains authoritative
- Deno and Notebook modules remain cold until Notebook execution or explicit preparation
- the namespace shape causes one intentional cold Code/Notebook tool-prefix transition; prewarm and live requests remain identical across stock, renamed, and configured proxy Responses Lite routes, while standard Responses stays flat

Closes #241
Closes #284
Closes #288
Refs #282

## Release

- Changeset: yes
- Package: `@howaboua/pi-codex-conversion`
- Aggregates: generated by release automation

### CODE REVIEW

**Comment by IgorWarzocha** [comment]
@codex please review this PR and give me 10-20 issues if any. Categorize findings as required, recommended, or optional.

File: packages/pi-codex-conversion/src/adapter/provider-request.ts [hunk 1]
```diff
@@ -30,6 +31,17 @@ function applyCodexRuntimePayload(payload: unknown, codeMode: boolean): unknown
...
 }
 
+export function rewriteCodexProviderHeaders(
+	headers: ProviderHeaders,
+	ctx: ExtensionContext,
+	state: AdapterState,
+): void {
+	if (state.config.voiceFeaturesOnly) return;
+	if (isCanonicalCodexAliasModel(ctx.model) && resolveCodexRuntimePlan(ctx, state.config).kind === "code") {
+		headers[RESPONSES_LITE_HEADER] = "true";
```
[the remark this answers is not in the document]
Comment (howaclawa) [reply packages/pi-codex-conversion/src/adapter/provider-request.ts]: Fixed in bb54e740 and bb7b7b56. Canonical alias endpoint trust is now credential-resolved once during before_agent_start, before prompt or provider transforms. A noncanonical resolved endpoint leaves the prompt, payload, and Responses Lite header untouched; the stored model-scoped decision is reused by both request hooks without a second auth resolution.

File: packages/pi-codex-conversion/tests/adapter-tool-activation.test.ts [hunk 2]
```diff
@@ -36,22 +40,52 @@ function createAdapterState(overrides: Partial<AdapterState["config"]> = {}): Ad
...
-function createContext(model: { provider: string; api: string; id: string; input?: string[] }, statuses?: unknown[]) {
+function createContext(model: { provider: string; api: string; id: string; baseUrl?: string; input?: string[] }, statuses?: unknown[]) {
 	return {
 		hasUI: Boolean(statuses),
 		model,
 		ui: { setStatus: (_key: string, value: unknown) => statuses?.push(value) },
 	};
 }
 
+test("Codex transport identity is broader than the canonical subscription trust boundary", () => {
```
[the remark this answers is not in the document]
Comment (howaclawa) [reply packages/pi-codex-conversion/tests/adapter-tool-activation.test.ts]: Fixed in bb54e740. Removed the six remaining net-new cases (15 complete cases culled from the supplied patch in total). The branch now has 73 permanent tests, exactly matching main, while the existing boundary cases retain the narrow routing assertions.

**Comment by IgorWarzocha** [comment]
@codex please review this PR and give me 10-20 issues if any. Categorize findings as required, recommended, or optional.
